pycode/memilio-epidata/memilio/epidata/assessNPIEffects.py

```diff
@@ -0,0 +1,595 @@
+    clusters_perp = [[] for i in range(idx_to_cluster_idx.max()+1)]
+    for ii in range(len(clusters)):
+        clusters_perp[ii] = list(indices_all.difference(set(clusters[ii])))
+
+    # old method
+    '''
+    # extract correlation values of block diagonals and offdiagonals separ.
+    corr_diag = []
+    corr_offdiag = []
+    for ii in range(len(clusters)):
+        corr_diag = np.append(corr_diag, abs(
+            corr_mat[np.ix_(clusters[ii], clusters[ii])].flatten()))
+        corr_offdiag = np.append(corr_offdiag, abs(
+            corr_mat[np.ix_(clusters[ii], clusters_perp[ii])].flatten()))
+
+    step_width=0.02
+    corr_thresholds = np.linspace(0+step_width, 1, int(1/step_width))
+    p_diag = np.array([len(np.where(corr_offdiag <= corr_thresholds[i])[0])
+                      for i in range(len(corr_thresholds))])/len(corr_offdiag)*step_width
+    p_offdiag = np.array([len(np.where(corr_diag < corr_thresholds[i])[0])
+                         for i in range(len(corr_thresholds))])/len(corr_diag)*step_width
+
+
+    return clusters, p_diag.sum()+p_offdiag.sum()
+    '''
+    # new method
+    if idx_to_cluster_idx.max() > 0:
+        sample_silhouette_values = silhouette_samples(
+            corr_mat, idx_to_cluster_idx)
+    else:
+        return clusters, np.nan
+
+    # get minimal cluster size with at least weak structuring (silhouette value 0 or greater than 0.25) in all clusters
+    if not ((sample_silhouette_values > 0.25) | (sample_silhouette_values == 0)).all():
+        return clusters, np.nan
+    else:
+        return clusters, idx_to_cluster_idx.max()+1
+
+# TODO: Used name 'methods' instead of 'metrics'. This is conform with documentation of hierarchy.linkage
+    # and does not lead to confusion with metric used in pdist. To discuss!
+
+
+def compute_hierarch_clustering(corr_pairwdist,
+                                methods=['single', 'complete', 'average',
+                                         'weighted', 'centroid', 'median',
+                                         'ward']):
+    """! Computes a hierarchical clustering for a (list of) method(s) and
+    provides the maximum cophenetic distance(s) as well as a score for the
+    clustering (see @method evaluate_clustering(...)).
+
+    @param corr_mat correlation matrix between the features / data set items
+        to be clustered hierarchically.
+    @param corr_pairwdist Computed pairwise distance between the features / data
+        set items.
+    @param method method or list of methods to compute the hierarchical
+        clustering.
+
+    @return (List of) hierarchical clustering(s), maximum cophenetic distance(s)
+        and scores of the hierarchical clustering.
+    """
+
+    # NOTE: if changing method, pay attention to linkage methods;
+    #       'centroid', 'median', and 'ward' are correctly defined only if
+    #       Euclidean pairwise metric is used in distance matrix that we used as input.
```

Comment:
Does this "still" hold? For ward, I don't see this at all directly, for the others, (only) the meaning changes